api/src/main/java/org/pmiops/workbench/db/dao/UserServiceImpl.java

```diff
@@ -582,46 +584,71 @@ public DbUser syncComplianceTrainingStatus(DbUser dbUser)
       return dbUser;
     }
 
-    Timestamp now = new Timestamp(clock.instant().toEpochMilli());
     try {
       Integer moodleId = dbUser.getMoodleId();
+      String email = dbUser.getUsername();
       if (moodleId == null) {
-        moodleId = complianceService.getMoodleId(dbUser.getUsername());
+        moodleId = complianceService.getMoodleId(email);
         if (moodleId == null) {
           // User has not yet created/logged into MOODLE
           return dbUser;
         }
         dbUser.setMoodleId(moodleId);
       }
 
-      List<BadgeDetails> badgeResponse = complianceService.getUserBadge(moodleId);
-      // The assumption here is that the User will always get 1 badge which will be AoU
-      if (badgeResponse != null && badgeResponse.size() > 0) {
-        BadgeDetails badge = badgeResponse.get(0);
-        Timestamp badgeExpiration =
-            badge.getDateexpire() == null
-                ? null
-                : new Timestamp(Long.parseLong(badge.getDateexpire()));
+      String expiryEpoch = null;
+      if (configProvider.get().featureFlags.enableMoodleV2Api) {
+        Map<String, BadgeDetails> userBadgesByName = complianceService.getUserBadgesByName(email);
+        if (userBadgesByName.containsKey(complianceService.getResearchEthicsTrainingField())) {
+          BadgeDetails badge =
+              userBadgesByName.get(complianceService.getResearchEthicsTrainingField());
+          expiryEpoch =
+              badge.getGlobalexpiration() == null
+                  ? badge.getDateexpire()
+                  : badge.getGlobalexpiration();
+        } else {
+          // Moodle has not returned research ethics training information for the given user --
+          // we should clear the user's training completion and expiration time.
+          dbUser.setComplianceTrainingCompletionTime(null);
+          dbUser.setComplianceTrainingExpirationTime(null);
+        }
+      } else {
+        List<BadgeDetailsDeprecated> badgeResponse = complianceService.getUserBadge(moodleId);
+        // The assumption here is that the User will always get 1 badge which will be AoU
+        if (badgeResponse != null && badgeResponse.size() > 0) {
+          BadgeDetailsDeprecated badge = badgeResponse.get(0);
+          expiryEpoch = badge.getDateexpire();
```

Comment:
Please make `expiryEpoch` an `Optional`, since it's not guaranteed to be there. You can do that with `badgeResponse.stream().findFirst()`, so you get rid of the scary `get(0)`.